### PR TITLE
Shutdown endpoint

### PR DESCRIPTION
**Pull request by brussee**
Includes fixes in session to gracefully shutdown.

### CODE REVIEW

**Comment by brussee** [comment]
@lfdversluis @devos50 Would one of you please review this please?

File: Tribler/Core/Modules/restapi/shutdown_endpoint.py [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+        """
+        gracetime = 2.0
+
+        def shutdown_process(_, code=1):
+            reactor.addSystemEventTrigger('after', 'shutdown', os._exit, code)
+            reactor.stop()
+            self.process_checker.remove_lock_file()
+
+        if not self._stopping:
+            self._stopping = True
```
Comment (brussee) [inline Tribler/Core/Modules/restapi/shutdown_endpoint.py]: @whirm @devos50  Is the endpoint instantiated for every request? If so, this ```if _stopping``` does not help.

**Comment by brussee** [comment]
test_anon_tunnel (Tribler.Test.test_tunnel_community_positive.TestTunnelCommunityPositive) ... Traceback (most recent call last):
  File "/home/jenkins/workspace/GH_Tribler_PR_tests_linux/tribler/Tribler/Main/tribler_main.py", line 150, in __init__
    session.start()
  File "/home/jenkins/workspace/GH_Tribler_PR_tests_linux/tribler/Tribler/Core/Session.py", line 466, in start
    startup_deferred = self.lm.register(self, self.sesslock)
  File "/home/jenkins/workspace/GH_Tribler_PR_tests_linux/tribler/Tribler/Core/APIImplementation/LaunchManyCore.py", line 114, in register
    self.metadata_store = LevelDbStore(self.session.get_metadata_store_dir())
  File "/home/jenkins/workspace/GH_Tribler_PR_tests_linux/tribler/Tribler/Core/leveldbstore.py", line 84, in __init__
    self._db = self._leveldb(os.path.relpath(store_dir, os.getcwdu()))
LevelDBError: IO error: lock collected_metadata/LOCK: already held by process
Traceback (most recent call last):
  File "/home/jenkins/workspace/GH_Tribler_PR_tests_linux/tribler/Tribler/Main/tribler_main.py", line 851, in OnExit
    torrent_db._db.clean_db(randint(0, 24) == 0, exiting=True)
AttributeError: 'NoneType' object has no attribute '_db'
ERROR

test_bad_request (Tribler.Test.Core.Modules.test_versioncheck.TestVersionCheck) ... + FAILURE=255

https://jenkins.tribler.org/job/GH_Tribler_PR_tests_linux/2734/artifact/output/test_runners_output/00003.err/*view*/

**Comment by brussee** [comment]
```
test_anon_download_without_exitnodes (Tribler.Test.test_tunnel_community_negative.TestTunnelCommunityNegative) ... Traceback (most recent call last):
  File "/home/jenkins/workspace/GH_Tribler_PR_tests_linux/tribler/Tribler/Main/tribler_main.py", line 150, in __init__
    session.start()
  File "/home/jenkins/workspace/GH_Tribler_PR_tests_linux/tribler/Tribler/Core/Session.py", line 466, in start
    startup_deferred = self.lm.register(self, self.sesslock)
  File "/home/jenkins/workspace/GH_Tribler_PR_tests_linux/tribler/Tribler/Core/APIImplementation/LaunchManyCore.py", line 114, in register
    self.metadata_store = LevelDbStore(self.session.get_metadata_store_dir())
  File "/home/jenkins/workspace/GH_Tribler_PR_tests_linux/tribler/Tribler/Core/leveldbstore.py", line 84, in __init__
    self._db = self._leveldb(os.path.relpath(store_dir, os.getcwdu()))
LevelDBError: IO error: lock collected_metadata/LOCK: already held by process
Traceback (most recent call last):
  File "/home/jenkins/workspace/GH_Tribler_PR_tests_linux/tribler/Tribler/Main/tribler_main.py", line 851, in OnExit
    torrent_db._db.clean_db(randint(0, 24) == 0, exiting=True)
AttributeError: 'NoneType' object has no attribute '_db'
ERROR
test_anon_download_without_relays (Tribler.Test.test_tunnel_community_negative.TestTunnelCommunityNegative) ... + FAILURE=255
```
https://jenkins.tribler.org/job/GH_Tribler_PR_tests_linux/2734/artifact/output/test_runners_output/00000.err/*view*/

**Comment by brussee** [comment]
retest this please

**Comment by brussee** [comment]
retest this please

**Comment by brussee** [comment]
@lfdversluis @devos50 Would one of you please review this please?

File: Tribler/Core/APIImplementation/LaunchManyCore.py [hunk 2]
```diff
@@ -680,29 +680,35 @@ def early_shutdown(self):
         self.shutdownstarttime = timemod.time()
         if self.boosting_manager:
             yield self.boosting_manager.shutdown()
-            self.boosting_manager = None
+        self.boosting_manager = None
```
Comment (lfdversluis) [inline Tribler/Core/APIImplementation/LaunchManyCore.py]: I don't think this is necessary, if `self.boosting_manager` fails it is already None. Same for the other cases.

File: Tribler/Core/APIImplementation/LaunchManyCore.py [hunk 3]
```diff
@@ -719,43 +725,50 @@ def early_shutdown(self):
...
+        if self.api_manager:
             yield self.api_manager.stop()
-            self.api_manager = None
+        self.api_manager = None
 
-        if self.watch_folder is not None:
+        if self.watch_folder:
             yield self.watch_folder.stop()
-            self.watch_folder = None
+        self.watch_folder = None
```
Comment (lfdversluis) [inline Tribler/Core/APIImplementation/LaunchManyCore.py]: Also, as you have added an endpoint too, I feel this should be 2 PRs. But if the others are ok with it, leave it as is.

File: Tribler/Core/Modules/restapi/shutdown_endpoint.py [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+            self.session.shutdown(gracetime=gracetime).addBoth(shutdown_process)
```
Comment (lfdversluis) [inline Tribler/Core/Modules/restapi/shutdown_endpoint.py]: you are swallowing any errors with addBoth, keep that in mind.

File: Tribler/Test/test_tunnel_community_negative.py [hunk 5]
```diff
@@ -86,6 +88,7 @@ def do_create_local_torrent(_):
 
         self.startTest(do_create_local_torrent, nr_exitnodes=5, nr_relays=0)
 
+    @skip("LevelDBError: IO error: lock collected_metadata/LOCK: already held by process")
```
Comment (lfdversluis) [inline Tribler/Test/test_tunnel_community_negative.py]: Have issues been created for these skips?

**Comment by lfdversluis** [comment]
`shutdown_process` is not covered by a test.

File: Tribler/Core/APIImplementation/LaunchManyCore.py (hunk 2 above)
Comment (brussee) [inline Tribler/Core/APIImplementation/LaunchManyCore.py]: For consistency with the other shutdown statements.

File: Tribler/Core/APIImplementation/LaunchManyCore.py (hunk 3 above)
Comment (brussee) [inline Tribler/Core/APIImplementation/LaunchManyCore.py]: These changes only came to light with adding the shutdown endpoint. Apparently nobody ever noticed any issue earlier.

File: Tribler/Core/Modules/restapi/shutdown_endpoint.py (hunk 4 above)
Comment (brussee) [inline Tribler/Core/Modules/restapi/shutdown_endpoint.py]: You need to shutdown, so any errors in the shutdown function that can be handled should be handled there (try-catch) any other errors that are unhandled should be logged here and then the system must shutdown.

File: Tribler/Test/test_tunnel_community_negative.py (hunk 5 above)
Comment (brussee) [inline Tribler/Test/test_tunnel_community_negative.py]: @devos50 just refactored them. I will rebase and retest.

**Comment by brussee** [comment]
That would require a custom test tear down function since you really shut the reactor down.

File: Tribler/Core/APIImplementation/LaunchManyCore.py (hunk 2 above)
Comment (lfdversluis) [inline Tribler/Core/APIImplementation/LaunchManyCore.py]: Then I'd argue the others need to be changed

Comment (brussee) [inline Tribler/Core/APIImplementation/LaunchManyCore.py]: Because anything can evaluate to False you have to set them to None nonetheless.

**Comment by brussee** [comment]
```
CRITICAL 1468929590         _legacy:twisted:154  Unhandled Error
Traceback (most recent call last):
  File "C:\Python27\lib\threading.py", line 763, in run
    self.__target(*self.__args, **self.__kwargs)
  File "C:\workspace\jenkins\workspace\GH_Tribler_PR_tests_win64\tribler\Tribler\Core\Utilities\twisted_thread.py", line 55, in _reactor_runner
    reactor.run(installSignalHandlers=False)
  File "C:\Python27\lib\site-packages\twisted\internet\base.py", line 1194, in run
    self.mainLoop()
  File "C:\Python27\lib\site-packages\twisted\internet\base.py", line 1203, in mainLoop
    self.runUntilCurrent()
--- <exception caught here> ---
  File "C:\Python27\lib\site-packages\twisted\internet\base.py", line 825, in runUntilCurrent
    call.func(*call.args, **call.kw)
  File "C:\workspace\jenkins\workspace\GH_Tribler_PR_tests_win64\tribler\Tribler\Core\Libtorrent\__init__.py", line 27, in <lambda>
    lambda_f = lambda a = args, kwa = kwargs: invoke_func(*a, **kwa)
  File "C:\workspace\jenkins\workspace\GH_Tribler_PR_tests_win64\tribler\Tribler\Core\Libtorrent\__init__.py", line 28, in invoke_func
    download.session.lm.threadpool.add_task(lambda_f, 1)
exceptions.AttributeError: 'NoneType' object has no attribute 'add_task'
```
https://jenkins.tribler.org/job/GH_Tribler_PR_tests_win64/2385/testReport/junit/Tribler.Test.Core.Modules.RestApi.test_downloads_endpoint/TestDownloadsEndpoint/test_start_download_hash_cache/

**Comment by brussee** [comment]
retest this please

**Comment by brussee** [comment]
@whirm Ready.

**Pull request merged**